crates/sui-core/src/state_accumulator.rs

```diff
@@ -0,0 +1,588 @@
+        self.watermarks
+            .insert(&CheckpointWatermark::HighestAccumulated, checkpoint_seq_num)
+    }
+
+    pub fn update_highest_checkpoint_ready_to_hash_seq_number(
+        &self,
+        checkpoint_seq_num: &CheckpointSequenceNumber,
+    ) -> Result<(), TypedStoreError> {
+        self.watermarks
+            .insert(&CheckpointWatermark::HighestReadyToHash, checkpoint_seq_num)
+    }
+}
+
+/// Thread safe clone-able handle for enqueueing items to Accumulator
+#[derive(Clone)]
+pub struct EnqueueHandle {
+    sender: mpsc::Sender<State>,
+    store: Arc<StateAccumulatorStore>,
+}
+
+impl EnqueueHandle {
+    /// Enqueue checkpoint state for accumulation. A checkpoint shall
+    /// be processed (accumulated both individually and towards
+    /// the root state hash) exactly once
+    ///
+    /// NOTE: When called from executor, this should be called before we increment
+    /// highest executed checkpoint watermark, otherwise we may inefficiently read effects
+    /// from the effects store rather than using effects passed in memory throuhg enqueue.
+    /// This may happen due to a race between scheduling new checkpoints for hashing and
+    /// checking the queue.
+    pub async fn enqueue(&self, state: State) -> SuiResult {
+        let seq_num = state.checkpoint_seq_num;
+
+        let highest = self.store.get_highest_checkpoint_ready_to_hash_seq_number().unwrap();
+
+        match highest {
+            // Check to ensure items are enqueued in order and no checkpoint sequence
+            // numbers are skipped
+            Some(highest) => {
+                match seq_num - highest {
+                    // next checkpoint in sequence - happy path
+                    1 => {},
+                    
+                    u64::MIN..=0 => {
+                        warn!("Checkpoint effects for sequence number {seq_num:?} already enqueued for accumulation");
+                        return Ok(());
+                    },
+                    
+                    2..=u64::MAX => panic!("Checkpoint seq num {seq_num:?} contents enqueued for accumulation out of order"),
+                }
+            }
+            None => {
+                if seq_num != 0 {
+                    panic!("Checkpoint seq num {seq_num:?} contents enqueued for accumulation out of order");
+                }
+            }
+        }
+
+        // note that we increment the watermark irrespective of whether `send` succeeds
+        // or not. This ensures that it will eventually be handled even in a crash
+        // recovery scenario
+        self.store.update_highest_checkpoint_ready_to_hash_seq_number(&seq_num)
+            .expect("StateAccumulator: failed to read highest cehckpoint ready to hash");
+
+        if state.end_of_epoch_flag {
+            // In this case we want to block until we have capacity to guarantee that this
+            // will eventually be accumulated
+            self.sender.send(state).await.map_err(|err| {
```

Comment:
if possible, since we're blocking anyway, could we just call a function directly here rather than using the channel? if nothing else it helps future readers of the code since they can use go-to-definition (or cmd-f).